folding-schemes/src/frontend/circom/browser.rs

```diff
@@ -0,0 +1,177 @@
+use crate::frontend::FCircuit;
+use crate::frontend::FpVar::Var;
+use crate::Error;
+use ark_circom::circom::{CircomCircuit, R1CS as CircomR1CS};
+use ark_circom::circom::{R1CSFile, R1CS};
+use ark_ff::{BigInteger, PrimeField};
+use ark_r1cs_std::alloc::AllocVar;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError};
+use ark_std::fmt::Debug;
+use byteorder::{LittleEndian, ReadBytesExt};
+use std::io::{BufReader, Cursor, Read};
+use std::usize;
+
+/// This circuit is the one we will use in order to fold from the browser.
+///
+/// A clear example on how to do all of this can be seen at:
+/// <https://github.com/CPerezz/wasm-sonobe-integration>
+#[derive(Clone, Debug)]
+pub struct CircomFCircuitBrowser<F: PrimeField> {
+    pub state_len: usize,
+    pub external_inputs_len: usize,
+    pub witness: Vec<F>,
+    pub r1cs: CircomR1CS<F>,
+}
+
+impl<F: PrimeField> FCircuit<F> for CircomFCircuitBrowser<F> {
+    /// (r1cs_file_bytes,  state_len, external_inputs_len)
+    type Params = (Vec<u8>, usize, usize);
+
+    fn new(params: Self::Params) -> Result<Self, Error> {
+        let (r1cs_bytes, state_len, external_inputs_len) = params;
+        let reader = BufReader::new(Cursor::new(&r1cs_bytes[..]));
+
+        let mut r1cs: R1CS<F> = R1CSFile::<F>::new(reader)
+            .expect("Error reading R1cs")
+            .into();
+
+        // That's some weird magic. Ask @dmpierre
+        r1cs.wire_mapping = None;
+
+        Ok(Self {
+            state_len,
+            external_inputs_len,
+            witness: vec![F::zero(); r1cs.num_variables],
+            r1cs,
+        })
+    }
+
+    fn state_len(&self) -> usize {
+        self.state_len
+    }
+    fn external_inputs_len(&self) -> usize {
+        self.external_inputs_len
+    }
+
+    fn step_native(
+        &self,
+        _i: usize,
+        z_i: Vec<F>,
+        external_inputs: Vec<F>,
+    ) -> Result<Vec<F>, Error> {
+        // Should we keep these?
+        assert_eq!(z_i.len(), self.state_len());
+        assert_eq!(external_inputs.len(), self.external_inputs_len());
+
+        // Extracts the z_i1(next state) from the witness vector and external inputs.
+        let z_i1 = z_i[..self.state_len()].to_vec();
```

Comment:
Can you explain to me why `z_i1` is `z_i[..self.state_len()]`? Why is it that the new state `z_i1` there if `z_i` isn't used anywhere here?

Reply:
This is the output of the IVC (witness passed from circom from the previous circuit execution).
And it's always in the first position, then we have the input and then the external inputs.

Reply:
Notice that this just works as the `CircomFCircuit` does. They do exactly the same just generating the witness in different places.

See: https://github.com/privacy-scaling-explorations/sonobe/blob/cb1b8e37aa07dda255d3f3651385a45613beed4e/folding-schemes/src/frontend/circom/mod.rs#L88

Reply:
Then the indexing is different compared to this

https://github.com/privacy-scaling-explorations/sonobe/blob/cb1b8e37aa07dda255d3f3651385a45613beed4e/folding-schemes/src/frontend/circom/mod.rs#L88 

I am wondering if this slice isn't including the constant 1 as well from the circom witness?

Reply:
That's true. I missed it!!!!! 

Thanks and good catch!!
fixed in: [5c67849](https://github.com/privacy-scaling-explorations/sonobe/pull/169/commits/5c678493efa6be5071316c5d484a08a30bf75bb1)